docs: add gas station docs (#5544)
iotaledger/iota: docs/content/operator/gas-station/architecture/components.mdx

---
title: Gas Station Components
description: Learn about the key components of the IOTA Gas Station
teams:
  - iotaledger/identity
tags:
  - gas-station
  - sponsored-transactions
  - explanation
  - reference
---

The **IOTA Gas Station** consists of several key components, each playing a crucial role in managing transactions efficiently and securely.

```mermaid
flowchart LR
 subgraph subGraph0["Gas Station Pool"]
        GSS["Gas Station Servers"]
  end
    KMS[("Key Management Store")] -.-> GSS
    GSS --> REDIS[("Persistent Storage (Redis)")]

    style KMS stroke-width:1px,stroke-dasharray: 5
```

##  Gas Station Server

The **Gas Station Server** is the central component of the system. Each deployment of a Gas Station Server is associated with a single sponsor wallet, responsible for funding transactions. To scale operations, multiple Gas Station Servers can be deployed together as a **Gas Station Pool**, allowing them to share the same sponsor address and persistent storage.

> ### HTTP Endpoints:

Reviewer: Consider linking the endpoints to the API description

Author: good point, fixed

    deactivate ReserveGas

    activate User
    User->>User: Prepare transaction & sign it
    deactivate User
    User->>ExecuteTx: Submit signed transaction
```

## Gas Station Storage

The **Gas Station Storage** serves as the persistent database for managing gas objects and transaction states. The storage layer, powered by Redis, keeps track of all available gas coins, ensuring that transactions can be sponsored efficiently. Redis together with LUA scripts provide a built-in mechanism that supports concurrent access from multiple Gas Station Servers, allowing for scalable and high-performance operations. This ensures that multiple servers in a Gas Station Pool can efficiently share and allocate gas resources without conflicts. It is essential for ensuring continuity, especially when gas pools are depleted and must be replenished.

To maintain consistency with the network, the storage state is periodically synchronized with the configured interval. This interval is defined in `config.yaml` under `refresh-interval-sec`, ensuring that the system regularly verifies and updates its stored state based on the actual conditions of the IOTA network.

> **Gas Station Storage usage process:**

Reviewer 2: ```suggestion
### Gas Station Storage Usage Process
```

* When switching between networks (e.g., from a testnet to a mainnet), all Redis entries should be cleared to prevent the Gas Station from returning outdated or non-existent objects. This can be done manually by flushing the Redis database before reinitializing the Gas Station instance. The storage layer, powered by Redis, keeps track of all available gas coins, ensuring that transactions can be sponsored efficiently.

:::

## Key Store Manager

The **Key Store Manager (KSM)** provides secure key management for signing transactions. It offers two configurations:

* **External Key Management Services (KMS)** – Uses external providers like AWS KMS for enhanced security.
* **In-memory key storage** – Maintains private keys within the Gas Station instance for faster but less secure operations.

**External Key Store** is an optional component that facilitates secure transaction signing by integrating an **external Key Management System (KMS)**. This feature simplifies key management, particularly when deploying multiple Gas Station Servers as a pool that shares the same signing key across multiple instances.

### External KMS Integration

> The Gas Station assumes that the external KMS will communicate using a **JSON-RPC API**. The KMS must expose the following two endpoints:

Reviewer: Is this really JSON-RPC?

Author: no, it's not :) leftover, fixed

* **GET `/get-pubkey-address`**

>   This endpoint returns the public key address corresponding to the stored private key encoded in base64

Reviewer 2: ```suggestion
  This endpoint returns the public key address corresponding to the stored private key encoded in base64.
```

* **POST `/sign-transaction`**

>   This endpoint is responsible for signing transactions. The request payload is a JSON object containing the base64-encoded and BCS-serialized transaction data

Reviewer 2: ```suggestion
  This endpoint is responsible for signing transactions. The request payload is a JSON object containing the base64-encoded and BCS-serialized transaction data.
```

```mermaid
sequenceDiagram
    participant Network as IOTA Network Node
    participant Gas_Station as Gas Station Server
    participant KMS as Key Management System (KMS)

    Note right of Gas_Station: On startup
    Gas_Station->>KMS: Request public key address (/get-pubkey-address)
    KMS-->>Gas_Station: Return public key address

    loop For each transaction
        Gas_Station->>KMS: Request signature (/sign-transaction)
        KMS-->>Gas_Station: Return signed transaction
        Gas_Station->>Network: Submit signed transaction

    end
```

> It is crucial that the serialized signature and keys  adhere to the **IOTA transaction signature specification**. Developers should refer to the [IOTA documentation](https://docs.iota.org/developer/cryptography/transaction-auth/signatures#signature-requirements) for further details on signature requirements.

Reviewer 2: ```suggestion
It is crucial that the serialized signature and keys  adhere to the **IOTA transaction signature specification**. Developers should refer to the [IOTA documentation](../../../developer/cryptography/transaction-auth/signatures.mdx#signature-requirements) for further details on signature requirements.
```
